#### Operaciones DataFrames

##### Definición de 2 DataFrames

In [1]:
import pandas as pd
import numpy as np

In [2]:
datos = [[7, 6, 3], [10, 1, 4], [4, 2, 5], [12, 5, 11]]
df1 = pd.DataFrame(datos, columns=["Col1", "Col2", "Col3"])
df1

Col1  Col2  Col3
0     7     6     3
1    10     1     4
2     4     2     5
3    12     5    11

In [3]:
filas = [
    ["A", "C", "B", "M", "P"],
    ["F", "K", "G", "Y", "E"],
    ["S", "D", "H", "Z", "W"],
]


columnas = ["Col2", "Col3", "Col4", "Col5", "Col6"]


dicc = dict(zip(columnas, zip(*filas)))


df2 = pd.DataFrame(dicc)


df2

Col2 Col3 Col4 Col5 Col6
0    A    C    B    M    P
1    F    K    G    Y    E
2    S    D    H    Z    W

#### Concat de concatenar

In [4]:
df3 = pd.concat([df1, df2], axis=0, join="outer")
df3

Col1 Col2 Col3 Col4 Col5 Col6
0   7.0    6    3  NaN  NaN  NaN
1  10.0    1    4  NaN  NaN  NaN
2   4.0    2    5  NaN  NaN  NaN
3  12.0    5   11  NaN  NaN  NaN
0   NaN    A    C    B    M    P
1   NaN    F    K    G    Y    E
2   NaN    S    D    H    Z    W

##### Leer un dataframe, head, tail, atributos, métodos.

In [5]:
df = pd.read_csv("Pesos_alturas_mundo.csv")
df

sexo                  País  Altura_media  Peso   IMC
0      m          Países Bajos          1.84  87.9  26.1
1      m            Montenegro          1.83  90.4  27.0
2      m               Estonia          1.82  89.9  27.0
3      m             Dinamarca          1.82  86.8  26.3
4      m  Bosnia y Hercegovina          1.82  87.1  26.4
..   ...                   ...           ...   ...   ...
243    f                 Nepal          1.52  51.6  22.3
244    f             Guatemala          1.51  61.9  27.3
245    f                 Yemen          1.54  56.9  24.0
246    f                  Laos          1.53  53.7  23.0
247    f        Timor Oriental          1.52  49.0  21.2

[248 rows x 5 columns]

In [6]:
df.head()

sexo                  País  Altura_media  Peso   IMC
0    m          Países Bajos          1.84  87.9  26.1
1    m            Montenegro          1.83  90.4  27.0
2    m               Estonia          1.82  89.9  27.0
3    m             Dinamarca          1.82  86.8  26.3
4    m  Bosnia y Hercegovina          1.82  87.1  26.4

In [7]:
df.tail()

sexo            País  Altura_media  Peso   IMC
243    f           Nepal          1.52  51.6  22.3
244    f       Guatemala          1.51  61.9  27.3
245    f           Yemen          1.54  56.9  24.0
246    f            Laos          1.53  53.7  23.0
247    f  Timor Oriental          1.52  49.0  21.2

In [8]:
df.shape

(248, 5)

In [9]:
df.values

array([['m', 'Países Bajos', 1.84, 87.9, 26.1],
       ['m', 'Montenegro', 1.83, 90.4, 27.0],
       ['m', 'Estonia', 1.82, 89.9, 27.0],
       ...,
       ['f', 'Yemen', 1.54, 56.9, 24.0],
       ['f', 'Laos', 1.53, 53.7, 23.0],
       ['f', 'Timor Oriental', 1.52, 49.0, 21.2]], dtype=object)

In [10]:
df.index

RangeIndex(start=0, stop=248, step=1)

In [11]:
df.columns

Index(['sexo', 'País', 'Altura_media', 'Peso', 'IMC'], dtype='object')

In [12]:
df.size

1240

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sexo          248 non-null    object 
 1   País          248 non-null    object 
 2   Altura_media  248 non-null    float64
 3   Peso          248 non-null    float64
 4   IMC           248 non-null    float64
dtypes: float64(3), object(2)
memory usage: 9.8+ KB


In [14]:
df.describe()

Altura_media        Peso         IMC
count    248.000000  248.000000  248.000000
mean       1.676774   74.999194   26.570565
std        0.079600   11.286519    2.765120
min        1.510000   49.000000   20.100000
25%        1.617500   68.375000   25.100000
50%        1.670000   74.550000   26.700000
75%        1.740000   84.000000   27.900000
max        1.840000  103.700000   35.200000

In [15]:
df.set_index("País")

sexo  Altura_media  Peso   IMC
País                                               
Países Bajos            m          1.84  87.9  26.1
Montenegro              m          1.83  90.4  27.0
Estonia                 m          1.82  89.9  27.0
Dinamarca               m          1.82  86.8  26.3
Bosnia y Hercegovina    m          1.82  87.1  26.4
...                   ...           ...   ...   ...
Nepal                   f          1.52  51.6  22.3
Guatemala               f          1.51  61.9  27.3
Yemen                   f          1.54  56.9  24.0
Laos                    f          1.53  53.7  23.0
Timor Oriental          f          1.52  49.0  21.2

[248 rows x 4 columns]

In [16]:
df.set_index("País", inplace=True)
df

sexo  Altura_media  Peso   IMC
País                                               
Países Bajos            m          1.84  87.9  26.1
Montenegro              m          1.83  90.4  27.0
Estonia                 m          1.82  89.9  27.0
Dinamarca               m          1.82  86.8  26.3
Bosnia y Hercegovina    m          1.82  87.1  26.4
...                   ...           ...   ...   ...
Nepal                   f          1.52  51.6  22.3
Guatemala               f          1.51  61.9  27.3
Yemen                   f          1.54  56.9  24.0
Laos                    f          1.53  53.7  23.0
Timor Oriental          f          1.52  49.0  21.2

[248 rows x 4 columns]

In [17]:
df.reset_index(inplace=True)
df

País sexo  Altura_media  Peso   IMC
0            Países Bajos    m          1.84  87.9  26.1
1              Montenegro    m          1.83  90.4  27.0
2                 Estonia    m          1.82  89.9  27.0
3               Dinamarca    m          1.82  86.8  26.3
4    Bosnia y Hercegovina    m          1.82  87.1  26.4
..                    ...  ...           ...   ...   ...
243                 Nepal    f          1.52  51.6  22.3
244             Guatemala    f          1.51  61.9  27.3
245                 Yemen    f          1.54  56.9  24.0
246                  Laos    f          1.53  53.7  23.0
247        Timor Oriental    f          1.52  49.0  21.2

[248 rows x 5 columns]

In [18]:
df = pd.read_csv("Pesos_alturas_mundo.csv", index_col=["País", "sexo"])
df.iloc[149:150, :2]  # df.iloc [fila,columna] en ese orden sino funciona mal.
df.iloc[[149, 150], [0, 2]]  # Solo fila 149 y 150 y altura y imc no pone el peso.
# df[149:150]['Peso'] # Slice solo funciona en filas no funciona para columnas
# df.loc['Australia',['Altura_media','Peso']] #df.loc[fila,columna] en ese orden sino no funciona
# df['Peso']['Australia'] #df[columna][fila] en ese orden sino no funciona

Altura_media   IMC
País      sexo                    
Australia f             1.65  26.8
Irlanda   f             1.64  27.1

In [19]:
df = pd.read_csv("Pesos_alturas_mundo.csv")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 248 entries, 0 to 247
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sexo          248 non-null    object 
 1   País          248 non-null    object 
 2   Altura_media  248 non-null    float64
 3   Peso          248 non-null    float64
 4   IMC           248 non-null    float64
dtypes: float64(3), object(2)
memory usage: 9.8+ KB


In [20]:
dicc = {i: range(7) for i in range(5)}
dff = pd.DataFrame(dicc).T
dff

0  1  2  3  4  5  6
0  0  1  2  3  4  5  6
1  0  1  2  3  4  5  6
2  0  1  2  3  4  5  6
3  0  1  2  3  4  5  6
4  0  1  2  3  4  5  6

##### Filtros

In [21]:
filtro = df["Altura_media"] >= 1.7
filtro2 = df.Altura_media >= 1.7
filtro3 = df.sexo == "f"

df[filtro2 & filtro3]

sexo          País  Altura_media  Peso   IMC
124    f  Países Bajos           1.7  73.2  25.3
125    f    Montenegro           1.7  75.3  26.2

##### Ordenamientos

In [22]:
df = pd.read_csv("Pesos_alturas_mundo.csv")
df.sort_values(by="Altura_media", ascending=False)

sexo                  País  Altura_media  Peso   IMC
0      m          Países Bajos          1.84  87.9  26.1
1      m            Montenegro          1.83  90.4  27.0
2      m               Estonia          1.82  89.9  27.0
3      m             Dinamarca          1.82  86.8  26.3
4      m  Bosnia y Hercegovina          1.82  87.1  26.4
..   ...                   ...           ...   ...   ...
246    f                  Laos          1.53  53.7  23.0
242    f            Bangladesh          1.52  50.5  21.8
243    f                 Nepal          1.52  51.6  22.3
247    f        Timor Oriental          1.52  49.0  21.2
244    f             Guatemala          1.51  61.9  27.3

[248 rows x 5 columns]

In [23]:
df.set_index(["sexo", "País"]).sort_index()[120:130]

Altura_media  Peso   IMC
sexo País                                       
f    Uruguay                    1.61  71.3  27.4
     Venezuela                  1.60  68.5  26.9
     Vietnam                    1.58  54.0  21.7
     Yemen                      1.54  56.9  24.0
m    Afganistán                 1.68  63.6  22.6
     Albania                    1.74  81.4  27.0
     Alemania                   1.80  88.7  27.4
     Andorra                    1.78  87.3  27.5
     Antigua y Barbuda          1.78  81.6  25.7
     Arabia Saudita             1.70  80.9  28.0

Supongamos que tienes un conjunto de datos que representa las calificaciones de estudiantes en diferentes materias. Crea un DataFrame con esta información y realiza algunas operaciones básicas:

Crea un DataFrame llamado calificaciones con las siguientes columnas: "Estudiante", "Materia", "Calificación".

Agrega al menos 10 filas de datos ficticios al DataFrame.

Muestra las primeras 5 filas del DataFrame.

Encuentra el promedio de calificaciones.

Encuentra la calificación más alta en cada materia.

Agrega una nueva columna llamada "Aprobado" que contenga True si la calificación es mayor o igual a 60, y False en caso contrario.

Guarda el DataFrame resultante en un archivo CSV llamado "calificaciones.csv".

Aquí tienes un punto de partida en código:

In [24]:
import pandas as pd
import numpy as np

# Paso 1: Crear DataFrame
np.random.seed(0)
data = {
    "Tienda": ["Tienda1", "Tienda2", "Tienda1", "Tienda2", "Tienda1", "Tienda2"],
    "Producto": ["ProdA", "ProdA", "ProdA", "ProdB", "ProdB", "ProdB"],
    "CantidadVendida": np.random.randint(1, 10, 6),
    "PrecioUnitario": np.random.uniform(10, 50, 6),
}
ventas = pd.DataFrame(data)
ventas

# Paso 2: Agregar la columna "Venta"

ventas["Venta"] = ventas["CantidadVendida"] * ventas["PrecioUnitario"]
ventas
# Paso 3: Encontrar el total de ventas por tienda
# ventasXTienda = ventas.groupby('Tienda')[['Venta']].sum()
# ventasXTienda

# Paso 4: Encontrar el producto más vendido por tienda
# vt2=ventas.groupby(['Tienda','Producto'])[['CantidadVendida']].sum().reset_index()
# vt2
# maximos=vt2.groupby('Tienda').CantidadVendida.idxmax()
# maximos
# vt2=vt2.loc[maximos][['Tienda','Producto']]
# vt2
# Paso 5: Calcular el monto total de ventas por producto
# ventasxProd=ventas.groupby('Producto')[['Venta']].sum()
# ventasxProd
# Paso 6: Encontrar la tienda con el mayor monto total de ventas
# tiendamax=ventas.groupby('Tienda')[['Venta']].sum().idxmax().values[0]
# tiendamax
# Paso 7: Calcular el precio promedio de cada producto
# res=ventas.groupby('Producto')[['PrecioUnitario']].mean()
# res =res.rename(columns={'PrecioUnitario':'PrecioPromedio'})
# res
# Paso 8: Guardar el DataFrame en un archivo CSV
# ventas.to_csv('analisis_ventas.csv', index=False)
# print("DataFrame guardado en 'analisis_ventas.csv'")

Tienda Producto  CantidadVendida  PrecioUnitario       Venta
0  Tienda1    ProdA                6       35.835765  215.014587
1  Tienda2    ProdA                1       27.503488   27.503488
2  Tienda1    ProdA                4       45.670920  182.683680
3  Tienda2    ProdB                4       48.546510  194.186042
4  Tienda1    ProdB                8       25.337661  202.701286
5  Tienda2    ProdB                4       41.669002  166.676006

#### Otro Ejercicio

In [25]:
import pandas as pd
import numpy as np

# Paso 1: Crear DataFrame
np.random.seed(0)
data = {
    "Tienda": ["Tienda1", "Tienda2", "Tienda1", "Tienda2", "Tienda1", "Tienda2"],
    "Producto": ["ProdB", "ProdB", "ProdA", "ProdB", "ProdA", "ProdA"],
    "Ventas": np.random.randint(10, 50, 6),
    "Devoluciones": np.random.randint(1, 10, 6),
    "PrecioUnitario": np.random.uniform(20, 100, 6),
}

df = pd.DataFrame(data)
df
# Paso 2: Agregar la columna "MontoNetoVentas"
df["MontoNetoVentas"] = (df.Ventas - df.Devoluciones) * df.PrecioUnitario
# df
# Paso 3: Encontrar el monto total de ventas netas por tienda
# ventaxTienda_df=df.groupby('Tienda')[['MontoNetoVentas']].sum()
# ventaxTienda_df
# Paso 4: Encontrar el producto con el precio unitario más alto por tienda
# ipumax=df.groupby('Tienda')['PrecioUnitario'].idxmax()
# ipumax
# df.iloc[ipumax.values,[0,1]]
# Paso 5: Encontrar la tienda con el mayor monto total de ventas netas
# tiendaNetod_df=df.groupby('Tienda')[['MontoNetoVentas']].sum()['MontoNetoVentas'].idxmax()
# tiendaNetod_df
# Paso 6: Calcular el porcentaje de devoluciones con respecto a las ventas por producto
# Mostrar el DataFrame con la nueva columna
# df['PorcDevol']=round(df['Devoluciones']*100/df['Ventas'],2)
# df

In [26]:
import pandas as pd
import numpy as np

np.random.seed(3)  # Para reproducibilidad
sz = 100
productos = ["ProdA", "ProdB", "ProdC", "ProdD"]
producto_precio = [[producto, np.random.uniform(20, 100)] for producto in productos]
indices = np.random.choice(len(producto_precio), size=sz)

data = {
    "Tienda": np.random.choice(["Tienda1", "Tienda2", "Tienda3"], size=sz),
    "Producto": [producto_precio[i][0] for i in indices],
    "Ventas": np.random.randint(10, 50, sz),
    "Devoluciones": np.random.randint(1, 10, sz),
    "PrecioUnitario": [producto_precio[i][1] for i in indices],
}

df = pd.DataFrame(data)

# Paso 2: Agregar la columna "VentasNetas"
df["VentasNetas"] = (df["Ventas"] - df["Devoluciones"]) * df["PrecioUnitario"]

# # Paso 3: Encontrar el monto total de ventas netas por tienda
# df.groupby('Tienda')[['VentasNetas']].sum()

# Paso 4: Encontrar el producto más vendido por tienda
# vpt_df=df.groupby(['Tienda','Producto'])[['VentasNetas']].sum().reset_index()
# imax=vpt_df.groupby('Tienda')['VentasNetas'].idxmax()
# vpt_df.iloc[imax]

# Paso 5: Encontrar el porcentaje de devoluciones con respecto a las ventas por producto
# Mostrar el DataFrame con la nueva columna
# df['pdevol']=100*df['Devoluciones']/df['Ventas']

# Paso 6: Encontrar la tienda con el mayor monto total de ventas netas
# df.groupby('Tienda')[['VentasNetas']].sum().idxmax().values[0]

# Paso 7: Encontrar el producto con la mayor cantidad total de devoluciones
# df.groupby('Producto')[['Devoluciones']].sum().idxmax().values[0]


# Paso 8: Calcular el promedio de ventas netas por producto
# prom_df=df.groupby('Producto')[['VentasNetas']].mean()
# prom_df=prom_df.rename(columns={'VentasNetas':'VentasNetasPromedio'})
# prom_df

# df.groupby('Producto')[['VentasNetas']].agg('mean')

In [27]:
import pandas as pd

data = {"A": [1, 2, 3], "B": [4, 5, 6], "C": [7, 8, 9], "D": ["aa", "bb", "cc"]}
df = pd.DataFrame(data)

# result = df.agg(['sum'])
# result = df.agg({'A': ['sum'], 'B': ['sum'], 'C': ['min']})
numeric_columns = df.select_dtypes(include="number")
numeric_columns
# df

A  B  C
0  1  4  7
1  2  5  8
2  3  6  9

##### **Ejercicio de Regresión Lineal Múltiple** (4 variables independientes)  
Dados 11 registros de propiedades (Tabla 1), con las características de ***Valor de la propiedad***, ***antigüedad***, ***accesos***, ***oficinas*** y ***superficie***.  
Determine a partir de una regresión lineal múltiple en que medida estas características pueden predecir el ***valor de la propiedad***.   
  
**Obtenga:**  
1) Los Coeficientes de regresion para cada variable y su desviación estandar.  
2) Error esperado en la prediccion del valor de la propiedad o intervalo de confianza.

**Ecuaciones:**  
  
$$ \begin{pmatrix}
y_1 \\
y_2 \\
\vdots \\
y_{n}\end{pmatrix}=
\begin{pmatrix}
1 & x_{1,1} & x_{1,2} & \cdots & x_{1,k} \\
1 & x_{2,1} & x_{2,2} & \cdots & x_{2,k} \\
\vdots  & \vdots  & \ddots & \vdots  \\
1 & x_{n,1} & x_{n,2} & \cdots & x_{n,k} 
\end{pmatrix}
\begin{pmatrix}
b_0 \\
b_1 \\
\vdots \\
b_k\end{pmatrix} +
\begin{pmatrix}
u_1 \\
u_2 \\
\vdots \\
u_n\end{pmatrix} \quad \text{en adelante} \quad y=Xb+u$$  
en donde   
$n=$ *número de registros de la muestra* (en este caso $n=11$)   
$k=$ *número de variables* independientes (en este caso $k=4$)  
$b=$ vector de coeficientes de la regresión  
$X=$ matriz de variables donde se agrega una columna de unos, ***"matriz de diseño"***.  
$u = $ vector de perturbaciones aleatorias  
**Una estimación para $y$ será entonces:**
$$\hat{y}=X\hat{b}$$
y para obtener esos $\hat{b}$ en regresión lineal múltiple se minimizan los cuadrados de los residuos, es decir la función:
$$(e^Te)=(y-Xb)^T(y-Xb)$$  
$$e=y-y(b)$$
para cierto $\hat{b}$ la función $(e^Te)=(y-Xb)^T(y-Xb)$ tendrá un mínimo.  
Entonces:  
$$\frac{\partial{e^Te}}{\partial{b}}=\frac{\partial}{\partial{b}}{(y-Xb)^T(y-Xb)}=\frac{\partial}{\partial{b}}\big({y^Ty-(Xb)^Ty-y^TXb+(Xb)^TXb}\big)$$  
$$\frac{\partial}{\partial{b}}\big({(Xb)^TXb-2y^TXb}\big)=2X^TXb-2y^TX$$  
$$X^TX\hat{b}-y^TX=0 \quad \implies \quad X^TX\hat{b}=y^TX \quad \implies \quad \hat{b}=(X^TX)^{-1}y^TX= (X^TX)^{-1}X^Ty$$  
$$\hat{b}={(X^TX)}^{-1}X^Ty \quad\text {son los coeficientes estimados de la regresión}$$  
**Ahora se puede calcular la varianza para los coeficientes estimados $\hat{b}$:**  
$$\hat{b}-b={(X^TX)}^{-1}X^Ty-b={(X^TX)}^{-1}X^T(Xb+u)-b=b+(X^TX)^{-1}X^Tu-b$$
$$(\hat{b}-b)=(X^TX)^{-1}X^Tu$$
$$(\hat{b}-b)^T=u^TX(X^TX)^{-1}$$
$$V_{\hat{b}}=E\big[(\hat{b}-b)^T(\hat{b}-b)\big]=E\big[(X^TX)^{-1}X^Tuu^TX(X^TX)^{-1}\big]=(X^TX)^{-1}X^TE\big[uu^T\big]X(X^TX)^{-1}$$  
$$E\big[uu^T\big]\quad\text{puede estimarse por}\quad v_{\hat{y}}=\frac{\hat{e}^T\hat{e}}{n-(k+1)}\quad\text{de modo que:}$$
$$V_{\hat{b}}=v_{\hat{y}}X(X^TX)^{-1}=\frac{\hat{e}^T\hat{e}}{n-(k+1)}X(X^TX)^{-1}$$
$V_{\hat{b}} \quad \text {es la matriz de varianzas covarianzas de los coeficientes de la regresión}$  
$E[\quad]\text{\quad es el valor esperado}$  
$\hat{e}=y-\hat{y}$  

**Resumen de Fórmulas básicas de la regresión lineal múltiple:**  
$$\hat{b}={(X^TX)}^{-1}X^Ty$$
$$\hat{e}=y-\hat{y}\quad\text{en donde}\quad \hat{y}=X\hat{b} \quad\text{son los valores ajustados}$$
$$ v_{\hat{y}} = \frac{\hat{e}^T\hat{e}}{n - (k+1)} \quad \sigma_{\hat{y}} = \sqrt{v_{\hat{y}}} \quad \text{es la desviación de los residuos.}$$
$$V_{\hat{b}} = v_{\hat{y}}(X^TX)^{-1}=v_{ij}$$ 
$$\sigma_D=\begin{pmatrix}\sqrt{v_{00}},&\sqrt{v_{11}},&\dots,\sqrt{v_{kk}}\end{pmatrix}\quad\text{son las desviaciones estandar de los coeficientes de la regresión.}$$

#### Presentación de los datos en el dataframe.

In [28]:
import numpy as np
import pandas as pd

data = pd.read_csv("ValorInmueble.csv")
df = pd.DataFrame(data)
df

Valor  Antigüedad  Accesos  Oficinas  Superficie
0   142.0          20      2.0         2        2310
1   144.0          12      2.0         2        2333
2   151.0          33      1.5         3        2356
3   360.0          40      2.0         3        2273
4   139.0          53      3.0         2        2402
5   169.0          23      2.0         4        2425
6   126.0          99      1.5         2        2448
7   142.9          34      2.0         2        2471
8   163.0          23      3.0         3        2494
9   189.0          55      4.0         4        2517
10  149.0          22      3.0         2        2540

#### Ejemplo de Regresión Lineal Múltiple

##### Datos

| Valor $y$ | Antigüedad $x_1$| Accesos $x_2$ | Oficinas $x_3$| Superficie $x_4$ |
|-------|------------|---------|----------|------------|
| 142.0 | 20         | 2.0     | 2        | 2310       |
| 144.0 | 12         | 2.0     | 2        | 2333       |
| 151.0 | 33         | 1.5     | 3        | 2356       |
| 360.0 | 40         | 2.0     | 3        | 2273       |
| 139.0 | 53         | 3.0     | 2        | 2402       |
| 169.0 | 23         | 2.0     | 4        | 2425       |
| 126.0 | 99         | 1.5     | 2        | 2448       |
| 142.9 | 34         | 2.0     | 2        | 2471       |
| 163.0 | 23         | 3.0     | 3        | 2494       |
| 189.0 | 55         | 4.0     | 4        | 2517       |
| 149.0 | 22         | 3.0     | 2        | 2540       |


##### Matrices y Vectores

Se construye la matriz **X** agregando una columna de "unos" (1) al principio:

$$
\begin{bmatrix}
1 & 20 & 2.0 & 2 & 2310 \\
1 & 12 & 2.0 & 2 & 2333 \\
1 & 33 & 1.5 & 3 & 2356 \\
1 & 40 & 2.0 & 3 & 2273 \\
1 & 53 & 3.0 & 2 & 2402 \\
1 & 23 & 2.0 & 4 & 2425 \\
1 & 99 & 1.5 & 2 & 2448 \\
1 & 34 & 2.0 & 2 & 2471 \\
1 & 23 & 3.0 & 3 & 2494 \\
1 & 55 & 4.0 & 4 & 2517 \\
1 & 22 & 3.0 & 2 & 2540 \\
\end{bmatrix}
$$

El vector de la variable dependiente **y** es:

$$ y =
\begin{bmatrix}
142.0 \\
144.0 \\
151.0 \\
360.0 \\
139.0 \\
169.0 \\
126.0 \\
142.9 \\
163.0 \\
189.0 \\
149.0 \\
\end{bmatrix}
$$

In [29]:
import numpy as np

x = np.array(
    [
        [1, 20, 2.0, 2, 2310],
        [1, 12, 2.0, 2, 2333],
        [1, 33, 1.5, 3, 2356],
        [1, 40, 2.0, 3, 2273],
        [1, 53, 3.0, 2, 2402],
        [1, 23, 2.0, 4, 2425],
        [1, 99, 1.5, 2, 2448],
        [1, 34, 2.0, 2, 2471],
        [1, 23, 3.0, 3, 2494],
        [1, 55, 4.0, 4, 2517],
        [1, 22, 3.0, 2, 2540],
    ]
)
k = len(x[0]) - 1

##### Ecuación Matricial

La ecuación matricial es:

$$ \begin{bmatrix}
142.0 \\
144.0 \\
151.0 \\
360.0 \\
139.0 \\
169.0 \\
126.0 \\
142.9 \\
163.0 \\
189.0 \\
149.0 \\
\end{bmatrix} = 
\begin{bmatrix}
1 & 20 & 2.0 & 2 & 2310 \\
1 & 12 & 2.0 & 2 & 2333 \\
1 & 33 & 1.5 & 3 & 2356 \\
1 & 40 & 2.0 & 3 & 2273 \\
1 & 53 & 3.0 & 2 & 2402 \\
1 & 23 & 2.0 & 4 & 2425 \\
1 & 99 & 1.5 & 2 & 2448 \\
1 & 34 & 2.0 & 2 & 2471 \\
1 & 23 & 3.0 & 3 & 2494 \\
1 & 55 & 4.0 & 4 & 2517 \\
1 & 22 & 3.0 & 2 & 2540 \\
\end{bmatrix}
·
\begin{bmatrix} b_0 \\ b_1 \\ b_2 \\ b_3 \\b_4 \end{bmatrix}
$$

##### Solución

La solución para el vector de coeficientes **b** se obtiene utilizando la fórmula:

$$ \hat{b} = (X^T X)^{-1} X^T y $$

Calcular $X^T X$:

$$
\begin{bmatrix}
11 & 414 & 26 & 29 & 26569 \\
414 & 21566 & 970 & 1080 & 1003988 \\
26 & 970 & 67.5 & 70.5 & 63206 \\
29 & 1080 & 70.5 & 83 & 70145 \\
26569 & 1003988 & 63206 & 70145 & 64251953 \\
\end{bmatrix}
$$

In [30]:
import numpy as np

xt = x.T  # Se calcula la traspuesta de X con el método .T
xtx = np.dot(
    xt, x
)  # Se calcula el producto entre XT y X es un producto matricial (producto punto).
xtx

array([[1.1000000e+01, 4.1400000e+02, 2.6000000e+01, 2.9000000e+01,
        2.6569000e+04],
       [4.1400000e+02, 2.1566000e+04, 9.7000000e+02, 1.0800000e+03,
        1.0039880e+06],
       [2.6000000e+01, 9.7000000e+02, 6.7500000e+01, 7.0500000e+01,
        6.3206000e+04],
       [2.9000000e+01, 1.0800000e+03, 7.0500000e+01, 8.3000000e+01,
        7.0145000e+04],
       [2.6569000e+04, 1.0039880e+06, 6.3206000e+04, 7.0145000e+04,
        6.4251953e+07]])

Calcular $(X^T X)^{-1}$:

$$
(\mathbf{X}^T \mathbf{X})^{-1} = 
\begin{bmatrix}
1.07665224 \times 10^2 & 2.88525398 \times 10^{-2} & 3.00540699  & -5.30286799 \times 10^{-1} & -4.73493426 \times 10^{-2} \\
2.88525398 \times 10^{-2} & 1.80180546 \times 10^{-4} & 1.28723649 \times 10^{-3} & 1.77174972 \times 10^{-4} & -1.62060637 \times 10^{-5} \\
3.00540699 & 1.28723649 \times 10^{-3} & 2.86286108 \times 10^{-1} & -6.07756694 \times 10^{-2} & -1.47816392 \times 10^{-3} \\
-5.30286799 \times 10^{-1} & 1.77174972 \times 10^{-4} & -6.07756694 \times 10^{-2} & 1.69855498 \times 10^{-1} & 9.08638756 \times 10^{-5} \\
-4.73493426 \times 10^{-2} & -1.62060637 \times 10^{-5} & -1.47816392 \times 10^{-3} & 9.08638756 \times 10^{-5} & 2.12032553 \times 10^{-5} \\
\end{bmatrix}
$$

In [31]:
xtx_1 = np.linalg.inv(
    xtx
)  # Se calcula la inversa del resultado anterior con el metodo .linalg.inv()


xtx_1

array([[ 1.07665224e+02,  2.88525398e-02,  3.00540699e+00,
        -5.30286799e-01, -4.73493426e-02],
       [ 2.88525398e-02,  1.80180546e-04,  1.28723649e-03,
         1.77174972e-04, -1.62060637e-05],
       [ 3.00540699e+00,  1.28723649e-03,  2.86286108e-01,
        -6.07756694e-02, -1.47816392e-03],
       [-5.30286799e-01,  1.77174972e-04, -6.07756694e-02,
         1.69855498e-01,  9.08638756e-05],
       [-4.73493426e-02, -1.62060637e-05, -1.47816392e-03,
         9.08638756e-05,  2.12032553e-05]])

Calcular $X^T y$:

$$ X^T y = 
\begin{bmatrix}
1874.9 \\
69959.6 \\
4440.3 \\
5139.8 \\
4503959.9 \\
\end{bmatrix}
$$

In [32]:
y = df["Valor"]  # Se establece "y" con los valores medidos.
n = len(y)
xty = np.dot(
    xt, y
)  # Se obtiene el producto de la traspuesta de X por "y" con el método .dot() (producto punto)
xty

array([1.8749000e+03, 6.9959600e+04, 4.4403000e+03, 5.1398000e+03,
       4.5039599e+06])

Multiplicar $(X^T X)^{-1}$ por $X^T y$ para obtener **b**:

$$ (X^T X)^{-1}·X^Ty = \hat{b} = 
\begin{bmatrix}
1239.8414 \\
0.3359 \\
26.1225 \\
30.5687 \\
-0.5069 \\
\end{bmatrix}
$$

In [33]:
b = np.dot(xtx_1, xty)
b

array([ 1.23984142e+03,  3.35884808e-01,  2.61225137e+01,  3.05687069e+01,
       -5.06909926e-01])

Se calcula ahora $\hat{e}=y-\hat{y}=y-X\hat{b}$:

In [34]:
ys = np.dot(x, b)
e = y - ys
e

0     -46.979632
1     -30.633625
2     -36.535728
3     114.978298
4     -40.550631
5     -23.830058
6      -6.499705
7      30.830479
8       9.592920
9     -20.187687
10     49.815368
Name: Valor, dtype: float64

Se calcula la varianza $v_{\hat{y}} = \frac{1}{n - (k+1)}\hat{e}^T\hat{e}=3981.08$:

In [58]:
et = e.T
vy = (1 / (n - k - 1)) * np.dot(et, e)
vy

3981.083467464503

Se calcula la desviación en y $\sigma_{\hat{y}}=\sqrt{v_{\hat{y}}}=63$:

In [36]:
sy = np.sqrt(vy)
sy

63.095827654960694

Ahora se calcula la matriz de varianza-covarianza de la regresión:
$$V_{\hat{b}} = v_{\hat{y}}(X^TX)^{-1}=v_{ij}$$ 

$$
\begin{bmatrix}
428624.245 & 114.864369 & 11964.7761 & -2111.11601 & -188.501685 \\
114.864369 & 0.717313791 & 5.12459589 & 0.705348352 & -0.0645176923 \\
11964.7761 & 5.12459589 & 1139.72889 & -241.953012 & -5.88469395 \\
-2111.11601 & 0.705348352 & -241.953012 & 676.208915 & 0.361736673 \\
-188.501685 & -0.0645176923 & -5.88469395 & 0.361736673 & 0.084411929
\end{bmatrix}
$$

In [37]:
vb = vy * xtx_1
vb

array([[ 4.28624245e+05,  1.14864369e+02,  1.19647761e+04,
        -2.11111601e+03, -1.88501685e+02],
       [ 1.14864369e+02,  7.17313791e-01,  5.12459589e+00,
         7.05348352e-01, -6.45176923e-02],
       [ 1.19647761e+04,  5.12459589e+00,  1.13972889e+03,
        -2.41953012e+02, -5.88469395e+00],
       [-2.11111601e+03,  7.05348352e-01, -2.41953012e+02,
         6.76208915e+02,  3.61736673e-01],
       [-1.88501685e+02, -6.45176923e-02, -5.88469395e+00,
         3.61736673e-01,  8.44119290e-02]])

Ahora se extrae la diagonal y se realiza la raiz cuadrada para obtener las desviaciones de los coeficientes de la regresión:
$$\sigma_D=\begin{pmatrix}\sqrt{v_{00}},&\sqrt{v_{11}},&\dots,\sqrt{v_{kk}}\end{pmatrix}$$
$$\sigma_D=
\begin{pmatrix}
654.6940, & 0.8469, & 33.7599, & 26.0040, & 0.2905 \\
\end{pmatrix}
$$

In [38]:
vd = np.diagonal(vb)
sd = np.sqrt(vd)
sd

array([6.54694009e+02, 8.46943794e-01, 3.37598710e+01, 2.60040173e+01,
       2.90537311e-01])

Se puede calcular también el coeficiente de determinación $R^2$:
$$R^2=1-\frac{(y-\hat y)^T(y-\hat y)}{(y-\bar y)^T(y-\bar y)}=1-\frac{\hat{e}^T\hat{e}}{\bar{e}^T\bar{e}}=0.4367$$
y el coeficiente de correlación múltiple, $\sqrt{R^2}=R$:  
$$R=0.6608$$

In [39]:
ym = df["Valor"].mean()
ym = np.full(11, ym)
em = y - ym
r2 = 1 - np.dot(et, e) / np.dot(em.T, em)
r2

0.43666872376417354

El coeficiente de determinación ajustado $R^2_{\text{ajustado}}$ se calcula utilizando la siguiente fórmula:

$$
R^2_{\text{ajustado}} = 1 - \frac{  (n - 1)}{(n - 1) - k}\cdot(1 - R^2)=0.0611
$$

Donde:
- $R^2$ es el coeficiente de determinación.
- $n$ es el número de observaciones.
- $k$ es el número de predictores (variables independientes).

Esta fórmula ajusta el $R^2$ para tener en cuenta el número de predictores en el modelo y el número total de observaciones. Cuanto mayor sea el $R^2_{\text{ajustado}}$, mejor se ajusta el modelo a los datos, considerando el número de predictores.


In [40]:
ra2 = 1 - (1 - r2) * (n - 1) / (n - k - 1)
ra2

0.0611145396069559

El estadístico F puede expresarse de la siguiente manera:

$$ F =\frac{MSR}{MSE}= \frac{(\bar Y-\hat Y)^T(\bar Y-\hat Y)/k}{(Y-\hat Y)^T(Y-\hat Y)/(n-k-1)} = 1.1627$$


Donde:
- MSR (Mean Square Regression) es la media de los cuadrados de la regresión.
- MSE (Mean Square Error) es la media de los cuadrados del error.
- $ Y $ es el vector de observaciones de la variable dependiente.
- $ \hat{Y} $ es el vector de valores predichos por el modelo de regresión.
- $ \bar{Y}$ es el vector cuyas componentes es la media de los $Y$.
- $ k $ es el número de parámetros en el modelo.
- $ n $ es el número de observaciones.

El estadístico F evalúa la significancia global del modelo de regresión. El numerador mide la reducción de la variabilidad explicada por el modelo, y el denominador mide la variabilidad no explicada por el modelo. La comparación entre estas dos cantidades proporciona información sobre si el modelo en su conjunto es estadísticamente significativo.

In [41]:
ysym = ys - ym
yys = y - ys
ve = np.dot(ysym.T, ysym) / 4
vne = np.dot(yys.T, yys) / (11 - 4 - 1)
F = ve / vne
F

1.1627316168615074

#### Coeficiente de Durbin-Watson:
$$dw=\frac{\displaystyle\sum_{i=2}^{n} (\hat{e}_i-\hat{e}_{i-1})^2}{\displaystyle\sum_{i=1}^{n} \hat{e}_i^2}$$  
En forma matricial:
$$dw=\frac{\hat{e}_r^T\hat{e}_r}{\hat{e}^T\hat{e}}$$  
$$dw=2.330$$
En donde:  
  
$\hat{e}_r=\hat{e}[1:]-\hat{e}[:-1]$  
$\hat{e}=y-\hat{y}$

In [42]:
er = e[1:].reset_index(drop=True) - e[:-1]
dw = np.dot(er.T, er) / np.dot(e.T, e)
dw

2.33016892195727

##### Kurtosis de los residuos  
$$K=\frac{\mu_4}{V^2}=\frac{\frac{1}{n}\displaystyle\sum_{i=1}^{n}(\hat{e}_i-\bar{\hat{e}})^4}{\Big(\frac{1}{n}\displaystyle\sum_{i=1}^{n}(\hat{e}_i-\bar{\hat{e}})^2\Big)^2}=\frac{\frac{1}{n}\displaystyle\sum_{i=1}^{n}E_i^4}{\Big(\frac{1}{n}\displaystyle\sum_{i=1}^{n}E_i^2\Big)^2}$$
$$K=n\frac{E^TE_DE_DE}{(E^TE)^2}=n\frac{E_2^TE_2}{(E^TE)^2}=3.7125$$

In [43]:
emn = e.mean()
E = e - emn
E2 = E**2
K = n * (np.dot(E2.T, E2) / np.dot(E.T, E) ** 2)
K

3.7125209308622495

##### Asimetría de la muestra (skew):
$$S=\frac{\frac{1}{n}\displaystyle\sum_{i=1}^{n}(\hat{e}_i-\bar{\hat{e}})^3}{\Big(\frac{1}{n}\displaystyle\sum_{i=1}^{n}(\hat{e}_i-\bar{\hat{e}})^2\Big)^{\frac{3}{2}}}=\frac{\frac{1}{n}\displaystyle\sum_{i=1}^{n}E_i^3}{\Big(\frac{1}{n}\displaystyle\sum_{i=1}^{n}E_i^2\Big)^{\frac{3}{2}}}$$
$$S=\sqrt{n}\frac{E^TE_DE}{(E^TE)^{\frac{3}{2}}}=\sqrt{n}\frac{E_2^TE}{(E^TE)^{\frac{3}{2}}}=1.2613$$

In [44]:
S = np.sqrt(n) * (np.dot(E2.T, E) / np.dot(E.T, E) ** (3 / 2))
S

1.2612600491455257

##### Jarque-Bera
$$JB=\frac{n}{6}\Big(S^2+\frac{1}{4}(K-3)^2\Big)=3.1591$$

In [45]:
JB = (n / 6) * (S**2 + (1 / 4) * (K - 3) ** 2)
JB

3.1491137897995882

##### Omnibus
$$OM=\frac{\frac{1}{4}(S^2+K^2)}{\frac{1}{6}S^2\frac{1}{24}K^2}=36\Bigg(\frac{S^2+K^2}{S^2K^2}\Bigg)$$

In [46]:
OM = 36 * ((S**2 + K**2) / (S**2 * K**2))
OM
S**2 / 6 + K**2 / 24

0.839413304515525

##### Distribución de Fisher-Snedecor
$$f(x,df1,df2)=\frac{\sqrt{\frac{(df1x)^{df1}df2^{df2}}{(df1x+df2)^{df1+df2}}}}{xBeta(\frac{df1}{2},\frac{df2}{2})}
=\frac{1}{Beta(\frac{df1}{2},\frac{df2}{2})}\Big(\frac{df1}{df2}\Big)^\frac{df1}{2}x^{(\frac{df1}{2}-1)}(1+\frac{df1}{df2}x)^{-(\frac{df1+df2}{2})}$$

$$P(F\leq x)=\frac{Beta(\frac{d_1x}{d_1x+d_2},\frac{d_1}{2},\frac{d_2}{2})}{Beta(1,\frac{d_1}{2},\frac{d_2}{2})}=I\Bigg(\frac{d_1x}{d_1x+d_2},\frac{d_1}{2},\frac{d_2}{2}\Bigg)$$
$$P(F\geq x)=1-P(F\leq x)$$
$Beta(z,a,b)=\int_0^z t^{a-1}(1-t)^{b-1} \mathrm{d}t$
$$P(F\leq x)=\frac{Beta(\frac{4 \cdot 1.163}{4 \cdot 1.163+6},\frac{4}{2},\frac{6}{2})}{Beta(1,\frac{4}{2},\frac{6}{2})}=0.587$$
$$P(F\geq x)=1-0.587=0.413$$

##### Función beta incompleta inferior regularizada:
$$I(x,a,b)=\frac{Beta(x,a,b)}{Beta(a,b)}$$
para el caso de Fisher-Snedecor:
$$P(F\geq x)=1-I\Bigg(\frac{d_1x}{d_1x+d_2},\frac{d_1}{2},\frac{d_2}{2}\Bigg)$$
para el caso de la t-Student:
$$P(t\geq x)=\frac{1}{2}I\Bigg(\frac{v}{v+t^2},\frac{v}{2},\frac{1}{2}\Bigg)$$
$$P(|t|\geq x)=I\Bigg(\frac{v}{v+t^2},\frac{v}{2},\frac{1}{2}\Bigg)$$


In [47]:
from scipy import integrate



def betaF(a, b, x=1):
    i, e = integrate.quad(lambda t: t ** (a - 1) * (1 - t) ** (b - 1), 0, x)
    return i



def FisherP(d1, d2, p):

    x = d1 * p / (d1 * p + d2)

    a, b = d1 / 2, d2 / 2

    fm = betaF(a, b, x) / betaF(a, b)

    return 1 - fm



FisherP(4, 6, 1.1627316168615074)

0.4129569048435814

In [48]:
def PStudent(dr, t):
    x = dr / (dr + t**2)
    a = dr / 2
    ps = betaF(a, 0.5, x) / betaF(a, 0.5)
    return ps


PStudent(6, 1.894)

0.10705673423352648

##### Inversa de la función de distribución acumulada de Student e inversa de la beta incompleta inferior:


##### Resumen:
El valor de una propiedad se puede estimar con la siguiente expresión lineal:
###### $$ \textbf{valor prop}(\pm63)=(1239\pm655)+(0.33\pm0.85)·\textbf{Antigüedad}+(26\pm34)·\textbf{Accesos}+(30\pm26)·\textbf{Oficinas}-(0.51\pm0.29)·\textbf{Superficie}$$

##### Biblioteca Statsmodels



In [49]:
import numpy as np
import statsmodels.api as sm

# Datos de entrada
x = np.array(
    [
        [1, 20, 2.0, 2, 2310],
        [1, 12, 2.0, 2, 2333],
        [1, 33, 1.5, 3, 2356],
        [1, 40, 2.0, 3, 2273],
        [1, 53, 3.0, 2, 2402],
        [1, 23, 2.0, 4, 2425],
        [1, 99, 1.5, 2, 2448],
        [1, 34, 2.0, 2, 2471],
        [1, 23, 3.0, 3, 2494],
        [1, 55, 4.0, 4, 2517],
        [1, 22, 3.0, 2, 2540],
    ]
)

y = np.array(
    [142.0, 144.0, 151.0, 360.0, 139.0, 169.0, 126.0, 142.9, 163.0, 189.0, 149.0]
)

# Ajuste de la regresión lineal múltiple
model = sm.OLS(y, x).fit()

# Imprimir los resultados
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.437
Model:                            OLS   Adj. R-squared:                  0.061
Method:                 Least Squares   F-statistic:                     1.163
Date:                Fri, 08 Dec 2023   Prob (F-statistic):              0.413
Time:                        15:22:19   Log-Likelihood:                -57.866
No. Observations:                  11   AIC:                             125.7
Df Residuals:                       6   BIC:                             127.7
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1239.8414    654.694      1.894      0.1

c:\Users\admin\Desktop\Icaro\Codigo\Resumen\.venv\Lib\site-packages\scipy\stats\_stats_py.py:1806: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


In [50]:
import numpy as np
from scipy.special import betainc, betaincinv


def inverse_beta_inc_interpolation(a, b, p, tol=1e-9, max_iter=100):
    # Establecer límites para la búsqueda
    lower_limit = 0.0
    upper_limit = 1.0

    # Realizar búsqueda binaria para encontrar un punto cercano a la solución
    for _ in range(max_iter):
        midpoint = (lower_limit + upper_limit) / 2.0
        cdf_midpoint = betainc(a, b, midpoint)

        if np.abs(cdf_midpoint - p) < tol:
            return midpoint

        if cdf_midpoint < p:
            lower_limit = midpoint
        else:
            upper_limit = midpoint

    # Si la búsqueda binaria no converge, usar el método de interpolación
    x = np.linspace(lower_limit, upper_limit, 1000)
    cdf_values = betainc(a, b, x)
    inverse_interp = np.interp(p, cdf_values, x)

    return inverse_interp


# Ejemplo de uso
a = 3.0
b = 0.5
p = 0.05

result = inverse_beta_inc_interpolation(a, b, p)
print("Inverse of Beta Incomplete Function:", result)

Inverse of Beta Incomplete Function: 0.500526487827301


In [51]:
pp = 0.500526487827301
np.sqrt((6 - 6 * pp) / pp)

2.446911846234417